File: attractor/pipeline/retry.py
"""Retry mechanism with backoff policies."""

from __future__ import annotations

import random

from pydantic import BaseModel

from ..exceptions import RateLimitError, ProviderError, AuthenticationError


class RetryPolicy(BaseModel):
    max_retries: int = 0
    initial_delay_ms: float = 200
    backoff_factor: float = 2.0
    max_delay_ms: float = 30000
    jitter: bool = True


# Preset policies
RETRY_NONE = RetryPolicy(max_retries=0)
RETRY_STANDARD = RetryPolicy(
    max_retries=4, initial_delay_ms=200, backoff_factor=2.0,
    max_delay_ms=30000, jitter=True,
)
RETRY_AGGRESSIVE = RetryPolicy(
    max_retries=4, initial_delay_ms=500, backoff_factor=2.0,
    max_delay_ms=60000, jitter=True,
)
RETRY_LINEAR = RetryPolicy(
    max_retries=2, initial_delay_ms=500, backoff_factor=1.0,
    max_delay_ms=500, jitter=False,
)
RETRY_PATIENT = RetryPolicy(
    max_retries=2, initial_delay_ms=2000, backoff_factor=2.0,
    max_delay_ms=30000, jitter=True,
)

PRESETS: dict[str, RetryPolicy] = {
    "none": RETRY_NONE,
    "standard": RETRY_STANDARD,
    "aggressive": RETRY_AGGRESSIVE,
    "linear": RETRY_LINEAR,
    "patient": RETRY_PATIENT,
}


def compute_delay(attempt: int, policy: RetryPolicy) -> float:
    """Compute delay in seconds for a retry attempt."""
    delay_ms = policy.initial_delay_ms * (policy.backoff_factor ** attempt)
    delay_ms = min(delay_ms, policy.max_delay_ms)
    if policy.jitter:
        delay_ms *= random.uniform(0.5, 1.5)
    return delay_ms / 1000.0


def is_retryable(error: Exception) -> bool:
    """Check if an error should trigger a retry."""
    if isinstance(error, RateLimitError):
        return True
    if isinstance(error, AuthenticationError):
        return False
    if isinstance(error, ProviderError):
        if error.status_code:
            # 4xx except 429 are not retryable
            if 400 <= error.status_code < 500 and error.status_code != 429:
                return False
            # 5xx are retryable
            if error.status_code >= 500:
                return True
        return False
    if isinstance(error, (ConnectionError, TimeoutError, OSError)):
        return True
    return False

File: internal/pipeline/retry.go
package pipeline

import (
	"errors"
	"math"
	"math/rand"
	"time"

	"github.com/nigelpepper/attractor/internal/aerr"
)

// RetryPolicy parameterizes backoff behavior.
type RetryPolicy struct {
	MaxRetries     int
	InitialDelayMS float64
	BackoffFactor  float64
	MaxDelayMS     float64
	Jitter         bool
}

// Preset retry policies.
var (
	RetryNone       = RetryPolicy{MaxRetries: 0}
	RetryStandard   = RetryPolicy{MaxRetries: 4, InitialDelayMS: 200, BackoffFactor: 2.0, MaxDelayMS: 30000, Jitter: true}
	RetryAggressive = RetryPolicy{MaxRetries: 4, InitialDelayMS: 500, BackoffFactor: 2.0, MaxDelayMS: 60000, Jitter: true}
	RetryLinear     = RetryPolicy{MaxRetries: 2, InitialDelayMS: 500, BackoffFactor: 1.0, MaxDelayMS: 500, Jitter: false}
	RetryPatient    = RetryPolicy{MaxRetries: 2, InitialDelayMS: 2000, BackoffFactor: 2.0, MaxDelayMS: 30000, Jitter: true}
)

// Presets maps preset names to policies.
var Presets = map[string]RetryPolicy{
	"none":       RetryNone,
	"standard":   RetryStandard,
	"aggressive": RetryAggressive,
	"linear":     RetryLinear,
	"patient":    RetryPatient,
}

// ComputeDelay returns the backoff delay for a retry attempt.
func ComputeDelay(attempt int, policy RetryPolicy) time.Duration {
	delayMS := policy.InitialDelayMS * math.Pow(policy.BackoffFactor, float64(attempt))
	if delayMS > policy.MaxDelayMS {
		delayMS = policy.MaxDelayMS
	}
	if policy.Jitter {
		delayMS *= 0.5 + rand.Float64()
	}
	return time.Duration(delayMS) * time.Millisecond
}

// IsRetryable reports whether an error should trigger a retry.
func IsRetryable(err error) bool {
	var rl *aerr.RateLimitError
	if errors.As(err, &rl) {
		return true
	}
	var auth *aerr.AuthenticationError
	if errors.As(err, &auth) {
		return false
	}
	var pe *aerr.ProviderError
	if errors.As(err, &pe) {
		if pe.StatusCode != 0 {
			if pe.StatusCode >= 400 && pe.StatusCode < 500 && pe.StatusCode != 429 {
				return false
			}
			if pe.StatusCode >= 500 {
				return true
			}
		}
		return false
	}
	var to *aerr.TimeoutError
	return errors.As(err, &to)
}
